### PR TITLE
initial implementation

### PR DESCRIPTION
**Pull request by notwaldorf**
Meet `iron-form`. It validates, serializes and submits any native elements or custom elements that have an `iron-form-element-behavior`. Please take a look!

//cc @morethanreal @cdata

### CODE REVIEW

**Comment by notwaldorf** [comment]
Woohoo! Now with actual tests! All the PRs this depended on have landed, so it's ready for review! :tada: :balloon: :dancer:

File: iron-form.html [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+
+    _handleFormError: function (event) {
+      this.fire('iron-form-error', event.detail);
+    },
+
+    _registerElement: function(e) {
+      this._customElements.push(e.target);
+    },
+
+    _serialize: function() {
```
Comment (morethanreal) [inline iron-form.html]: Do you think this should be public? It seems like a useful function.

**Comment by morethanreal** [comment]
lgtm!

File: iron-form.html (hunk 1 above)
Comment (notwaldorf) [inline iron-form.html]: Sure! Done.

**Comment by notwaldorf** [comment]
@cdata does the ajax stuff look ok?

File: iron-form.html [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+    },
+
+    listeners: {
+      'iron-form-element-register': '_registerElement',
+      'submit': 'submit'
+    },
+
+    ready: function() {
+      // Object that handles the ajax form submission request.
+      this._requestBot = document.createElement('iron-ajax');
```
Comment (cdata) [inline iron-form.html]: :space_invader:

File: iron-form.html [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    },
+
+    /**
+     * Called to submit the form.
+     */
+    submit: function(event) {
+      var valid = true;
+
+      for (var el, i = 0; el = this._customElements[i], i < this._customElements.length; i++) {
+        valid = el.validate() && valid;
```
Comment (cdata) [inline iron-form.html]: This (lines 78 - 84) seems like a good candidate to be factored into a `isValid` method for the form. What do you think?

File: iron-form.html [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+      // Also go through the form's native elements.
+      for (var el, i = 0; el = this.elements[i], i < this.elements.length; i++) {
+        // Don't re-add custom elements that extend native elements (like an
+        // `<input is="fancy-input">`), since they will also show up in this list.
+        if (el.name && !json[el.name]) {
+          json[el.name] = el.value;
+        }
+      }
+
+      this.jsonData = json;
```
Comment (cdata) [inline iron-form.html]: It seems potentially weird that `jsonData` is a persistent property of the form, but its values are likely to fall out-of-sync with the values of the actual form inputs. Is it useful or necessary for `jsonData` to be a property?

File: iron-form.html [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+      if (this.method == 'POST') {
+        this._requestBot.body = JSON.stringify(this.jsonData);
+      }
+
+      this._requestBot.generateRequest();
+      this.fire('iron-form-submit');
+
+      // Don't perform a page refresh.
+      if (event)
```
Comment (cdata) [inline iron-form.html]: Minor style detail, but this conditional should probably have opening and closing `{ ... }`

File: test/basic.html [hunk 6]
```diff
@@ -0,0 +1,171 @@
+<!doctype html>
+<!--
```
Comment (cdata) [inline test/basic.html]: `@license`

File: test/simple-element.html [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+@license
+Copyright (c) 2015 The Polymer Project Authors. All rights reserved.
+This code may only be used under the BSD style license found at http://polymer.github.io/LICENSE.txt
+The complete set of authors may be found at http://polymer.github.io/AUTHORS.txt
+The complete set of contributors may be found at http://polymer.github.io/CONTRIBUTORS.txt
+Code distributed by Google as part of the polymer project is also
+subject to an additional IP rights grant found at http://polymer.github.io/PATENTS.txt
+-->
+<link rel="import" href="../../polymer/polymer.html">
+<link rel="import" href="../../iron-form-element-behavior/iron-form-element-behavior.html">
```
Comment (cdata) [inline test/simple-element.html]: Not actionable in this PR, but regarding the name `iron-form-element-behavior`, `element` and `behavior` seem like exclusive qualifications.

Comment (cdata) [inline test/simple-element.html]: Oh, I just got what this is. Sorry. Name is still.. confusing :)

**Comment by cdata** [comment]
Sorry, I didn't realise you weren't soliciting a full review until it was too late. The ajax-specific work looks fine.

File: iron-form.html (hunk 4 above)
Comment (notwaldorf) [inline iron-form.html]: It's probably not super useful. I'll make `serialize` return a json instead.

File: test/simple-element.html (hunk 7 above)
Comment (notwaldorf) [inline test/simple-element.html]: Yeah, I don't have a better solution :(

But if you do, i'll take it!

File: iron-form.html (hunk 3 above)
Comment (notwaldorf) [inline iron-form.html]: :+1:

**Pull request merged**